crates/driver/src/domain/competition/solution/encoding.rs

```diff
@@ -7,3 +22,330 @@ pub enum Strategy {
     /// Use logic from this module for encoding
     Domain,
 }
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("invalid interaction: {0:?}")]
+    InvalidInteractionExecution(competition::solution::interaction::Liquidity),
+    #[error("invalid clearing price: {0:?}")]
+    InvalidClearingPrice(eth::TokenAddress),
+    #[error(transparent)]
+    Math(#[from] Math),
+}
+
+pub fn tx(
+    auction_id: competition::auction::Id,
+    solution: &super::Solution,
+    contract: &contracts::GPv2Settlement,
+    approvals: impl Iterator<Item = eth::allowance::Approval>,
+    internalization: settlement::Internalization,
+) -> Result<eth::Tx, Error> {
+    let mut tokens = Vec::new();
+    let mut clearing_prices = Vec::new();
+    let mut trades: Vec<Trade> = Vec::new();
+    let mut pre_interactions = Vec::new();
+    let mut interactions = Vec::new();
+    let mut post_interactions = Vec::new();
+
+    // Encode uniform clearing price vector
+    for (token, price) in solution.prices.clone() {
+        tokens.push(token.0.into());
```

Comment:
It shouldn't be needed to access the internal type of `TokenAddress`

```suggestion
        tokens.push(token.into());
```